pkg/driver/controller.go

```diff
@@ -1129,3 +1133,32 @@ func LabelIndexFunc(label string) cache.IndexFunc {
 		return vs, nil
 	}
 }
+
+func (cs *controller) validateDeleteSnapshotReq(req *csi.DeleteSnapshotRequest) error {
+	if req.GetSnapshotId() == "" {
+		return status.Errorf(codes.InvalidArgument, "DeleteSnapshot: empty snapshotID")
+	}
+
+	cloneList, err := zfs.GetClonesForSnapshot(req.SnapshotId)
+	if err != nil {
+		return status.Errorf(
+			codes.NotFound,
+			"failed to handle delete snapshot request for {%s}, "+
+				"validation failed checking for existing clones. Error: %s",
+			req.GetSnapshotId(),
+			err.Error(),
+		)
+	}
+
+	// snapshot delete is not supported if clones exist for this snapshot
+	if len(cloneList.Items) != 0 {
+		return status.Errorf(
+			codes.Internal,
```

Comment:
Should be FAILED_PRECONDITION

Reply:
I think we should first evaluate the issue with what happens if we delete the volume/snapshot.

Reply:
verified that if we delete the volume i.e pv and pvc  the underneath zfs snapshot get deleted. Even though the zfs snapshot cr is available. To rectify this i have raised the PR https://github.com/openebs/zfs-localpv/pull/607

Before deleting the pvc
```
root@node-0-290533:~# zfs list -t all
NAME                                                                                                USED  AVAIL     REFER  MOUNTPOINT
zfspv-pool                                                                                          186K  9.20G       24K  /zfspv-pool
zfspv-pool/pvc-dd7c6d67-8451-43c3-bc56-993e9f41b1b3                                                24.5K  4.00G     24.5K  legacy
zfspv-pool/pvc-dd7c6d67-8451-43c3-bc56-993e9f41b1b3@snapshot-cc55c781-c4d5-4f7b-ac84-1b9582b469c6     0B      -     24.5K  -
zfspv-pool/pvc-dd7c6d67-8451-43c3-bc56-993e9f41b1b3@snapshot-19b7e7d1-f81f-435f-ae88-e73aa828ced8     0B      -     24.5K  -
```
After deleting the pvc
```
root@node-0-290533:~# zfs list -t all
NAME         USED  AVAIL     REFER  MOUNTPOINT
zfspv-pool   141K  9.20G       24K  /zfspv-pool
```